## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [165]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [166]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [167]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [168]:
# Use this space to do any additional research
# and familiarize yourself with the data.
# Load the dataset

# Get the shape of the dataset
print("Dataset shape:", df.shape)

# Get the data types of each column
print("Data types:\n", df.dtypes)

# Get the number of missing values in each column
print("Missing values:\n", df.isnull().sum())

Dataset shape: (54639, 18)
Data types:
 first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object
Missing values:
 first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64


In [169]:
# What three item categories had the most entries?
# Count the number of entries in each category
category_counts = df['category'].value_counts()

# Get the top 3 categories with the most entries
top_categories = category_counts.nlargest(3)

print(top_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [170]:
# For the category with the most entries, which subcategory had the most entries?

# Find the category with the most entries
max_category = df['category'].value_counts().idxmax()

# Find the most common subcategory within the top category
max_subcategory = df[df['category'] == max_category]['subcategory'].value_counts().idxmax()

print(f"Category with the most entries: {max_category}")
print(f"Subcategory with the most entries in {max_category}: {max_subcategory}")

Category with the most entries: consumables
Subcategory with the most entries in consumables: bathroom supplies


In [171]:
# Which five clients had the most entries in the data?

# Count the number of entries for each client
client_entries = df['client_id'].value_counts()

# Get the top 5 clients with the most entries
top_5_clients = client_entries.nlargest(5)

print(top_5_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [172]:
# Store the client ids of those top 5 clients in a list.
top_5_client_ids = top_5_clients.index.tolist()
print(top_5_client_ids)

[33615, 66037, 46820, 38378, 24741]


In [173]:
# How many total units (the qty column) did the client with the most entries order order?
top_client = top_5_client_ids[0]  # First client in the list
#Total units ordered by this beast
top_client_total_units = df.loc[df['client_id'] == top_client, 'qty'].sum()
print(f"The top client with the most entries: {top_client} ordered a total of {top_client_total_units} units.")

The top client with the most entries: 33615 ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [174]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
print(df[['unit_price', 'qty', 'line_subtotal']].head())

   unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00


In [175]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
print(df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head())

   unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20


In [176]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['line_price'] = ((df['line_subtotal'] + df['shipping_price']) * 1.0925).round(2)

print(df[['line_subtotal', 'shipping_price', 'line_price']].head())

   line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12
3        1056.18          356.70     1543.57
4        3902.00         6500.20    11364.40


In [177]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty'] + df['shipping_price'])
print(df[['unit_cost', 'qty', 'shipping_price', 'line_cost']].head())


   unit_cost  qty  shipping_price  line_cost
0     762.71  105         5512.50   85597.05
1      15.09   21          312.90     629.79
2       7.86   39          458.64     765.18
3      24.85   29          356.70    1077.35
4     108.17   20         6500.20    8663.60


In [178]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = (df['line_price'] - df['line_cost'])
print(df[['line_price', 'line_cost', 'line_profit']].head())


   line_price  line_cost  line_profit
0   131839.08   85597.05     46242.03
1      914.26     629.79       284.47
2     1077.12     765.18       311.94
3     1543.57    1077.35       466.22
4    11364.40    8663.60      2700.80


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [179]:
# Check your work using the totals above
order_totals = df.groupby('order_id')['line_price'].sum()

In [180]:
# Check your work using the totals above
print("Order 2742071: $", order_totals[2742071])
print("Order 2173913: $", order_totals[2173913])
print("Order 6128929: $", order_totals[6128929])

Order 2742071: $ 152811.88999999998
Order 2173913: $ 162388.71
Order 6128929: $ 923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [181]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_clients_spending = df.groupby('client_id')['line_price'].sum()
print("Total spending for the top 5 clients by quantity:")
for client_id in top_5_client_ids:
    amount_spent = top_clients_spending.get(client_id, "Client ID not found")
    print(f"Client {client_id}: ${amount_spent}")

Total spending for the top 5 clients by quantity:
Client 33615: $8377308.52
Client 66037: $10259514.790000001
Client 46820: $9743794.36
Client 38378: $12906550.870000001
Client 24741: $82268892.02


In [182]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

# Filter only for the top 5 clients
filtered_df = df[df['client_id'].isin(top_5_client_ids)]

# Sum the required metrics using .sum()
summary_df = filtered_df.groupby('client_id')[['qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']].sum().round(2)

#Reset index for readability
summary_df = summary_df.reset_index()

# Print the summary DataFrame
print(summary_df)

   client_id     qty  shipping_price   line_price    line_cost  line_profit
0      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
1      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
3      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
4      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81


In [183]:
# Format the data and rename the columns to names suitable for presentation.
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['shipping_price', 'line_price', 'line_cost', 'line_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
    """Convert a dollar amount to millions and format as a string."""
    return f"${value / 1_000_000:.2f}M"

# Apply the currency_format_millions function to only the money columns. 
for col in money_columns:
    summary_df[col] = summary_df[col].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
summary_df.rename(columns={
    'qty': 'Units',
    'shipping_price': 'Shipping (Millions)',
    'line_price': 'Revenue (Millions)',
    'line_cost': 'Cost (Millions)',
    'line_profit': 'Profit (Millions)'
}, inplace=True)

# Print the formatted DataFrame
print(summary_df)



   client_id   Units Shipping (Millions) Revenue (Millions) Cost (Millions)  \
0      24741  239862              $5.13M            $82.27M         $45.69M   
1      33615   64313              $1.83M             $8.38M          $6.18M   
2      38378   73667              $3.43M            $12.91M          $9.63M   
3      46820   75768              $1.60M             $9.74M          $7.01M   
4      66037   43018              $1.40M            $10.26M          $7.00M   

  Profit (Millions)  
0           $36.58M  
1            $2.20M  
2            $3.27M  
3            $2.74M  
4            $3.26M  


In [184]:
# Sort the updated data by "Profit (Millions)" from highest to lowest.
sorted_summary_df = summary_df.sort_values(by='Profit (Millions)', ascending=False)

# Assign the sorted data to a new DataFrame.
sorted_summary_df = sorted_summary_df.reset_index(drop=True)

# Print the sorted DataFrame.
print(sorted_summary_df)


   client_id   Units Shipping (Millions) Revenue (Millions) Cost (Millions)  \
0      24741  239862              $5.13M            $82.27M         $45.69M   
1      38378   73667              $3.43M            $12.91M          $9.63M   
2      66037   43018              $1.40M            $10.26M          $7.00M   
3      46820   75768              $1.60M             $9.74M          $7.01M   
4      33615   64313              $1.83M             $8.38M          $6.18M   

  Profit (Millions)  
0           $36.58M  
1            $3.27M  
2            $3.26M  
3            $2.74M  
4            $2.20M  


#Findings
Clients that ordered the most in quantity of units did not always translate to the highest profit or of other quantities. differences in shipping weights was more likely a factor. Over all there was a large variance and without graphing out in pandas, correlations and causations are harder to determine.